### PR TITLE
Removed push_back/push_front/insert extensions which took no parameters.

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
The call sites have been replaced with calls to emplace_back or emplace_front.

Corrected return types for AZStd insert() functions to return an `iterator` instance for every insert function which accepts a `const_iterator` position parameter as many of those functions were incorrectly returning void.

Updated AZStd::deque emplace_back/emplace_front function to correctly return a reference instead of void.
Updated AZStd::ring_buffer extension to support emplace_front and emplace_back functions.
Updated AZStd::forward_list to support rvalues and emplace_front/emplace_front functions. AZStd::forward_list previously didn't support any rvalue functions for move.

Since AZStd::forward_list is only used in unit test it is not much of a problem.
It is also has a lot of limitations as compared to AZStd::list which makes it not quite useful to ever use in the codebase.

Finally removed the swap() function extensions which accepted an rvalue reference.
They have been replaced with assignment to a default constructed instance, which what the swap function was swapping with it anyway.
The operation to swap with an rvalue is really just a move operation since it is stealing the values from a temporary.

Signed-off-by: lumberyard-employee-dm <56135373+lumberyard-employee-dm@users.noreply.github.com>

## How was this PR tested?

Ran automated test suite.

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/std/containers/deque.h [hunk 1]
```diff
@@ -474,13 +474,13 @@ namespace AZStd
...
         {
-            iterator last, first = begin();
+            iterator first = begin();
             if (insertPos == first)
             {
                 push_front(value);
                 return begin();
             }
-            else if (insertPos == (last = end()))
+            else if (iterator last = end(); insertPos == last)
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/std/containers/deque.h:483]: Nice update! 🙂

File: Code/Framework/AzCore/AzCore/std/containers/forward_list.h [hunk 2]
```diff
@@ -131,13 +132,12 @@ namespace AZStd
 
 
     /**
-    * The list container (double linked list) is complaint with \ref CStd (23.2.2). In addition we introduce the following \ref ListExtensions "extensions".
+    * The list container (single linked list) is complaint with \ref CStd (23.2.2). In addition we introduce the following \ref ListExtensions "extensions".
```
Comment (hultonha) [inline Code/Framework/AzCore/AzCore/std/containers/forward_list.h:135]: Lol what a typo 😅

**Review by hultonha — APPROVED** [review]
Looks like a great set of changes! 😄 Only a couple of small suggestions

**Comment by rgba16f** [comment]
Is there a corresponding PR for o3de-atom-sampleviewer?

File: Code/Tools/AssetProcessor/native/AssetDatabase/AssetDatabase.cpp [hunk 3]
```diff
@@ -1496,8 +1496,7 @@ namespace AssetProcessor
             [&](ScanFolderDatabaseEntry& scanFolder)
             {
                 found = true;
-                container.push_back();
-                container.back() = AZStd::move(scanFolder);
+                container.emplace_back() = AZStd::move(scanFolder);
```
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Tools/AssetProcessor/native/AssetDatabase/AssetDatabase.cpp:1499]: Yes this is safe.

The `QueryScanFoldersTable` function passes in lvalue reference to a ScanFolderDatabaseEntry which can be moved just fine.
Now I don't know how if the scanFolder is re-used after it is moved, but that is up to the AssetDatabaseConnection API to determine if it is passing it's parameters to a function by lvalue reference or rvalue reference

File: Code/Framework/AzCore/AzCore/Serialization/EditContext.h [hunk 4]
```diff
@@ -556,7 +555,7 @@ namespace AZ
     {
         if (IsValid())
         {
-            m_classElement->m_elements.push_back();
+            m_classElement->m_elements.emplace_back();
             Edit::ElementData& ed = m_classElement->m_elements.back();
```
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Serialization/EditContext.h:559]: Yes it can.
I missed consolidating this change.

**Comment by lumberyard-employee-dm** [comment]
> Is there a corresponding PR for o3de-atom-sampleviewer?

There is one now at https://github.com/o3de/o3de-atom-sampleviewer/pull/480

**Review by lawsonamzn — APPROVED** [review]

**Pull request merged**